### ISSUE TITLE
Why are some constants like MPI_REQUEST_NULL statics ?

### ISSUE THREAD

**Issue by gnzlbg** [bsteinb/rsmpi#62]
I've noticed that some constants like `MPI_REQUEST_NULL`, `MPI_COMM_WORLD`, etc. are `static mut`s instead of `const`s. Why is this ? In particular, when can I not read their values because it is undefined behavior to do so ? (e.g. before initializing MPI ?)

`static mut` is almost impossible to use correctly - https://github.com/rust-lang/rust/issues/53639

**Comment by gnzlbg** [comment]
E.g. `MPI_REQUEST_NULL` in:

* `mpich`: https://github.com/pmodels/mpich/blob/45fc41144de975f5bf938c87c036e2231166f269/src/include/mpi.h.in#L93
   
   ```c
    #define MPI_REQUEST_NULL   ((MPI_Request)0x2c000000)
   ```

* `openmpi`: https://github.com/open-mpi/ompi/blob/c79a4c55a24fba10edfbfa2d5a75ac11e950021e/ompi/include/mpi.h.in#L754

   ```c
   #define MPI_REQUEST_NULL OMPI_PREDEFINED_GLOBAL(MPI_Request, ompi_request_null)
   ```

is always a `const`.

**Comment by AndrewGaspar** [comment]
Actually, in this example you showed, OpenMPI's MPI_REQUEST_NULL is NOT a const. It's an extern. Here's the precise define for `ompi_request_null`, which is what `MPI_REQUEST_NULL` evaluates to after some casts:

```c
OMPI_DECLSPEC extern struct ompi_predefined_request_t ompi_request_null;
```

So unfortunately we can't make them const since the MPI standard does not require that they evaluate to constant expressions.

**Comment by gnzlbg** [comment]
Damn. Does this mean that the value of these is allowed to change during a program?

**Comment by AndrewGaspar** [comment]
I think they're defined once the program has been linked, but OpenMPI especially supports dynamic linking, so you don't necessarily know that value until runtime.

**Comment by gnzlbg** [comment]
Shouldn't then `static` (as opposed to `static mut`) be enough ? Or am I allowed to mutate these values ?

**Comment by AndrewGaspar** [comment]
Not sure - you'd have to ask the OpenMPI guys. Anything under `ffi` is a pedantic translation from bindgen. I think we generally treat usage of `ffi` as "at your own caution" (it DOES require `unsafe` everywhere).

**Comment by bsteinb** [comment]
The MPI specification 3.1 goes into details in section 2.5.4. Most MPI constants are link-time constants. I agree, that they should be `static mut`. In `rsmpi.h`, I have made them `extern const` which the version of `bindgen` that we have to use translates into `static mut`. More recent versions of `bindgen` use `static` (immutable) instead. Once again, I have to point to #44.

**Comment by bsteinb** [comment]
Fixed in a3adac0499e273b407c0040f0da5d30c1fb223c8? What used to be `static mut`s now seem to be `static`.
